<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Benjamin A Taylor" />


<title>Brassicas_2021_RNAseq</title>

<script src="Brassicas_2021_rnaseq_files/header-attrs-2.9/header-attrs.js"></script>
<script src="Brassicas_2021_rnaseq_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Brassicas_2021_rnaseq_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="Brassicas_2021_rnaseq_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Brassicas_2021_rnaseq_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Brassicas_2021_rnaseq_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Brassicas_2021_rnaseq_files/navigation-1.1/tabsets.js"></script>
<script src="Brassicas_2021_rnaseq_files/navigation-1.1/codefolding.js"></script>
<link href="Brassicas_2021_rnaseq_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="Brassicas_2021_rnaseq_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>




</head>

<body>


<div class="container-fluid main-container">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Brassicas_2021_RNAseq</h1>
<h4 class="author">Benjamin A Taylor</h4>
<h4 class="date">15/09/2021</h4>

</div>


<div id="data-import-pre-filtering-and-qc" class="section level2">
<h2>Data import, pre-filtering and QC</h2>
<p>We begin by performing QC on the RNAseq data, removing genes with low expression and samples that have evidence of being mislabeled. First for Brassica:</p>
<pre><code>## [1] &quot;13595/48645 Brassica genes filtered due to very low expression.&quot;</code></pre>
<pre><code>## [1] &quot;Removing an additional 104 Brassica genes with many low counts.&quot;</code></pre>
<p>Looking at the PCA, we can see a strange cluster of outlying Brassicas. These are all B. rapa tricoloris; we won’t exclude them, but we’ll take a note of their labels in case we want to try re-running analyses without them later. <img src="Analysis/RNAseq/Images/brass_clean_pca.png" width="1476" /></p>
<p>Now repeat QC and pre-filtering for Raphanus samples:</p>
<pre><code>## [1] &quot;14461/53691 Raphanus genes filtered due to very low expression.&quot;</code></pre>
<pre><code>## [1] &quot;Removing an additional 605 genes with many low counts.&quot;</code></pre>
<p>We see unusual clustering in the PCA for Raphanus, as with Brassica. The ‘outlying’ samples on PC2 are represent all Asian sativus var. caudatus, while the U.S. caudatus samples cluster with the other wilds. As with the unusual B. tricoloris samples, we’ll retain these for now but make a note in case we wish to exclude them at a later time. <img src="Analysis/RNAseq/Images/raph_clean_pca.png" width="1476" /></p>
</div>
<div id="parental-effect-handling" class="section level2">
<h2>Parental effect handling</h2>
<p>Before we proceed with the analysis, we need to decide how to handle the fact that a subset of samples have been standardised for parental effects, while the remaining samples have not received such a standardisation. Checking how many genes exhibit a difference between standardised and un-standardised conditions under liberal conditions (p&lt;0.1, lfc threshold=0) will give us some indication of how important this standardisation is.</p>
<p>Note that for Raphanus, all species/subspecies have both standardised and unstandardised samples, so we can just compare all against all, whereas for Brassica standardisation is present only for some B rapa (both wild and domesticated) so we need to subset before running the DESeq2 comparison.</p>
<pre><code>## [1] &quot;Number of Brassica rapa genes with parental effects at p&lt;0.1: 173/34946&quot;</code></pre>
<pre><code>## [1] &quot;Number of GO terms enriched among Brassica rapa parental DEGs: 26&quot;</code></pre>
<pre><code>## [1] &quot;Number of Raphanus genes with parental effects at p&lt;0.1: 85/38625&quot;</code></pre>
<pre><code>## [1] &quot;Number of GO terms enriched among Raphanus parental DEGs: 26&quot;</code></pre>
<p>Just out of interest, we’d also like to know whether the fucntions associated with parental effects in each case are the same- in fact, there’s no overlap:</p>
<pre><code>## [1] &quot;Number of shared GO terms enriched among parental effect DEGs in Brassica and Raphanus: 0 (hypergeometric p=1).&quot;</code></pre>
<p>The number of genes that display parental effeects in each case is quite small. At this stage we could choose to exclude these genes entirely, or we could retain them but keep tabs as we go along to see if these genes overlap strongly with other sets of genes that we pull out as interesting. For now we’ll go with the latter option.</p>
</div>
<div id="work-package-1-comparing-domesticates-to-their-wild-progenitors" class="section level2">
<h2>Work package 1: Comparing domesticates to their wild progenitors</h2>
<p>For this work package, we wish to compare wild progenitor species to modern domesticates. For Brassica, the comparison is between domesticated B rapa with wild B rapa, while for Raphanus the comparison is between R raphanistrum (wild) and all other Raphanus samples, all of which appear to be domesticated. Our analytic design for DESeq2 is expression ~ domestication + treatment + domestication*treatment.</p>
<p>First, let’s run the model for Brassica and check that the fit looks good:</p>
<pre><code>## [1] &quot;Check for gene expression outliers&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-brassica-1.png" width="960" /><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-brassica-2.png" width="960" /></p>
<pre><code>## [1] &quot;Local fit:&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-brassica-3.png" width="960" /></p>
<pre><code>## [1] &quot;Parametric fit&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-brassica-4.png" width="960" /></p>
<p>There are no apparent gene expression outliers. The parametric model is smoother and levels off more soundly to the right of the x axis, so we’ll move forward using this model for the results (remember, in general it’s better to overestimate than underestimate the dispersion parameter for genes with low expression).</p>
<p>Now the results for Brassica:</p>
<pre><code>## [1] &quot;Genes DE with treatment in Brassica rapa: 9 upregulated in no-competition treatment; 16 upregulated in competition treatment.&quot;</code></pre>
<pre><code>## [1] &quot;Genes DE with domestication status in Brassica rapa: 2073 upregulated in domesticated samples; 1818 upregulated in wild samples.&quot;</code></pre>
<pre><code>## [1] &quot;Genes DE with interaction between treatment and domestication in Brassica rapa: 41.&quot;</code></pre>
<pre><code>##                   DEGs GO_terms
## Domesticated_bias 2073       55
## Wild_bias         1818       85
## Unstressed_bias      9       15
## Stressed_bias       16       13
## Interaction         41       25</code></pre>
<p>We’re particularly interested in the interaction DEGs, since these are genes that, under domestication, have expressed a change in the plasticity of their responses to a cultivated (i.e. no-competition) environment. Interaction terms are hard to interpret, but plotting helps. Here’s the 12 most significant interaction terms:</p>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-results-brassica-interactionplots-1.png" width="960" /></p>
<p>There doesn’t appear to be an overall trend in terms of gains or losses of plasticity, but we can check this by asking whether the fold changes on average are higher or lower for wild or domesticated samples:</p>
<p>We must therefore conclude that Brassica interaction genes do not exhibit a consistent overall change in plasticity between wild and domesticated accessions.</p>
<p>Now, let’s repeat the above analyses using Raphanus. First check that the fit looks good:</p>
<pre><code>## [1] &quot;Check for gene expression outliers&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-raphanus-1.png" width="960" /><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-raphanus-2.png" width="960" /></p>
<pre><code>## [1] &quot;Local fit:&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-raphanus-3.png" width="960" /></p>
<pre><code>## [1] &quot;Parametric fit&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-model-raphanus-4.png" width="960" /></p>
<p>There are no apparent gene expression outliers. The parametric model is smoother and levels off more soundly to the right of the x axis, so we’ll move forward using this model for the results (remember, in general it’s better to overestimate than underestimate the dispersion parameter for genes with low expression).</p>
<pre><code>## [1] &quot;Genes DE with treatment in Raphanus: 76 upregulated in no-competition treatment; 53 upregulated in competition treatment.&quot;</code></pre>
<pre><code>## [1] &quot;Genes DE with domestication status in Raphanus: 533 upregulated in domesticated samples; 179 upregulated in wild samples.&quot;</code></pre>
<pre><code>## [1] &quot;Genes DE with interaction between treatment and domestication in Raphanus: 98.&quot;</code></pre>
<pre><code>##                   DEGs GO_terms
## Domesticated_bias  533       48
## Wild_bias          179       64
## Unstressed_bias     76       34
## Stressed_bias       53       27
## Interaction         98       49</code></pre>
<p>Plotting the 12 most significant interaction terms for Raphanus:</p>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-domesticate-results-raphanus-interactionplots-1.png" width="960" /></p>
<p>Are Raphanus interaction genes statistically of greater plasticity in wild or domesticated samples?</p>
<pre><code>## 
## FALSE  TRUE 
##    29    69</code></pre>
<pre><code>## [1] &quot;Probability that changes in plasticity of Raphanus genes during domestication are non-random in their directionality:&quot;</code></pre>
<pre><code>## [1] &quot;chi-squared p=5.33e-05; paired t-test p=0.15&quot;</code></pre>
<p>At least based on the chi-squared output, it appears that Raphanus genes on average have increased in plasticity during domestication. However, this result might be misleading, because the values for the Raphanus domesticates combine multiple different species, which may have resulted in depressed estimates of gene expression plasticity. For a more robust analysis, we should re-run the above for each Raphanus domesticate individually:</p>
<pre><code>##                          species up_wild up_dom    chisq      t       wilcox
## 1    Raphanus raphanistrum munra      39     59 0.043400 0.0484 0.0008654630
## 2               Raphanus sativus      31     67 0.000276 0.0928 0.0005876102
## 3 Raphanus sativus var. caudatus      32     66 0.000594 0.7170 0.0002252579</code></pre>
<p>When subsetting by species, the results are more interesting: per a chi-squared test, interaction genes are significantly more likely to be up-regulated than down-regulated. Paired t-tests gives less consistent results, but paired Wilcoxon tests concur with the chi-squared results. We can therefore tentatively conclude that in Raphanus, genes with significant interactions between domestication and treatment are significantly more like than chance to have increased in their gene expression plasticity in the course of domestication: in other words, humans appear to have selected for increased gene expression plasticity in radishes.</p>
</div>
<div id="work-package-2-comparing-wild-progenitors-to-never-domesticated-wild-species" class="section level2">
<h2>Work package 2: Comparing wild progenitors to never-domesticated wild species</h2>
<p>We now wish to ask a different question: do ‘progenitor species’ (i.e. those that were the original targets of human selection) differ systematically in their patterns of gene expression from wild species that were never domesticated? We are particularly interested in knowing whether progenitor species express greater gene expression plasticity than their never-domesticated relatives, since this could suggest that earty human farmers either intentionally or inadvertently targeted species that were particularly receptive to the benefits of cultivation.</p>
<p>Unfortunately, for Raphanus we have no data for non-domesticated wild species, so this analysis will be restricted to Brassica, comparing wild Brassica rapa to a number of never-domesticated Brassicas. We’ll begin by fitting a DESeq2 model as before, except that instead of comparing wild to domesticated, we’ll compare progenitors to non-progenitors.</p>
<pre><code>## [1] &quot;Check for gene expression outliers&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-wilds-model-brassica-1.png" width="960" /><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-wilds-model-brassica-2.png" width="960" /></p>
<pre><code>## [1] &quot;Local fit:&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-wilds-model-brassica-3.png" width="960" /></p>
<pre><code>## [1] &quot;Local fit:&quot;</code></pre>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-wilds-model-brassica-4.png" width="960" /></p>
<p>Results summary:</p>
<pre><code>##                   DEGs GO_terms
## Domesticated_bias 8710       53
## Wild_bias         9520       75
## Unstressed_bias      9       40
## Stressed_bias       11       10
## Interaction         63       30</code></pre>
<p>Plot top 12 interaction terms:</p>
<p><img src="Brassicas_2021_rnaseq_files/figure-html/deseq-wilds-results-brassica-interactionplots-1.png" width="960" /></p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
