<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Benjamin A Taylor" />


<title>Brassicas_2021_summarized_results</title>

<script src="Brassicas_2021_results_summary_files/header-attrs-2.9/header-attrs.js"></script>
<script src="Brassicas_2021_results_summary_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Brassicas_2021_results_summary_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="Brassicas_2021_results_summary_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Brassicas_2021_results_summary_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Brassicas_2021_results_summary_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Brassicas_2021_results_summary_files/navigation-1.1/tabsets.js"></script>
<script src="Brassicas_2021_results_summary_files/navigation-1.1/codefolding.js"></script>
<link href="Brassicas_2021_results_summary_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="Brassicas_2021_results_summary_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>




</head>

<body>


<div class="container-fluid main-container">




<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Brassicas_2021_summarized_results</h1>
<h4 class="author">Benjamin A Taylor</h4>
<h4 class="date">25/09/2021</h4>

</div>


<p>This markdown document summarizes the principal results of the Brassica project generation 2 analyses. Analyses are not performed within this document; prior to running this file, the results should be generated by running the two other .Rmd files in this repository: “Brassicas_2021_morphology.Rmd” and “Brassicas_2021_rnaseq.Rmd”.</p>
<div id="work-package-1-comparing-domesticates-to-their-wild-progenitors" class="section level2">
<h2>Work package 1: Comparing domesticates to their wild progenitors</h2>
<p>For this work package, we wish to compare wild progenitor species to modern domesticates. For Brassica, the comparison is between domesticated B rapa with wild B rapa, while for Raphanus the comparison is between R raphanistrum (wild) and all other Raphanus samples, all of which appear to be domesticated.</p>
<div id="brassica-gene-expression" class="section level3">
<h3>Brassica gene expression</h3>
<p>For the gene expression analysis, out model design is:</p>
<p><span class="math display">\[gene expression \sim domestication + treatment + domestication*treatment\]</span> Results summary (number of genes and GO terms enriched for each dependent variable) for wild vs domesticated Brassica rapa:</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">DEGs</th>
<th align="right">GO_terms</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Domesticated_bias</td>
<td align="right">2073</td>
<td align="right">55</td>
</tr>
<tr class="even">
<td>Wild_bias</td>
<td align="right">1818</td>
<td align="right">85</td>
</tr>
<tr class="odd">
<td>Unstressed_bias</td>
<td align="right">9</td>
<td align="right">15</td>
</tr>
<tr class="even">
<td>Stressed_bias</td>
<td align="right">16</td>
<td align="right">13</td>
</tr>
<tr class="odd">
<td>Interaction</td>
<td align="right">41</td>
<td align="right">25</td>
</tr>
</tbody>
</table>
<p>We can see from this that a great many genes are associated with domestication, but few with stress. 41 genes show an interaction between treatment and domestication, and are therefore of particular interest. Summary of the 25 GO terms associated with these interaction genes:</p>
<p><img src="Analysis/RNAseq/Images/brassica_domesticate_vs_progenitor_GO_treeplot.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We can also plot these interaction terms to get a better idea of what the reaction norms look like for wild and domesticates accessions respectively:</p>
<p><img src="Analysis/RNAseq/Images/brapas_interaction_deg_plots.png" width="2362" /></p>
<p>Eyeballing these plots, there’s no obvious trend in terms of increases or decreases in plasticity for wild vs domesticated accessions. Let’s run statistical tests to check whether this is actually the case:</p>
<pre><code>## [1] &quot;Interaction genes with higher plasticity in wild: 17&quot;</code></pre>
<pre><code>## [1] &quot;Interaction genes with higher plasticity in domesticate: 24&quot;</code></pre>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">Chi.squared</th>
<th align="right">Paired.Wilcoxon</th>
<th align="right">Paired.t.test</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p-value</td>
<td align="right">0.274</td>
<td align="right">0.221</td>
<td align="right">0.786</td>
</tr>
</tbody>
</table>
<p>From this we must conclude that, in Brassica rapa, genes have not changed systematically in their plasticity in the course of domestication.</p>
</div>
<div id="raphanus-gene-expression" class="section level3">
<h3>Raphanus gene expression</h3>
<p>Results summary (number of genes and GO terms enriched for each dependent variable) for wild vs domesticated Raphanus:</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">DEGs</th>
<th align="right">GO_terms</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Domesticated_bias</td>
<td align="right">533</td>
<td align="right">48</td>
</tr>
<tr class="even">
<td>Wild_bias</td>
<td align="right">179</td>
<td align="right">64</td>
</tr>
<tr class="odd">
<td>Unstressed_bias</td>
<td align="right">76</td>
<td align="right">34</td>
</tr>
<tr class="even">
<td>Stressed_bias</td>
<td align="right">53</td>
<td align="right">27</td>
</tr>
<tr class="odd">
<td>Interaction</td>
<td align="right">98</td>
<td align="right">49</td>
</tr>
</tbody>
</table>
<p>Compared to Brassica, we find that fewer genes are associated with domestication, but more with stress. 98 genes show an interaction between treatment and domestication, and are therefore of particular interest. Summary of the 49 GO terms associated with these interaction genes:</p>
<p><img src="Analysis/RNAseq/Images/raphanus_domesticate_vs_progenitor_GO_treeplot.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We can also plot these interaction terms to get a better idea of what the reaction norms look like for wild and domesticates accessions respectively:</p>
<p><img src="Analysis/RNAseq/Images/raphanus_interaction_deg_plots.png" width="2362" /></p>
<p>Eyeballing these plots, it looks like wild accessions tend to exhibit higher gene expression plasticity than domesticated accessions. Let’s run statistical tests to check whether this is actually the case:</p>
<pre><code>## [1] &quot;Interaction genes with higher plasticity in wild: 69&quot;</code></pre>
<pre><code>## [1] &quot;Interaction genes with higher plasticity in domesticate: 29&quot;</code></pre>
<pre><code>## [1] &quot;Plasticity higher in wild vs domesticate?:&quot;</code></pre>
<pre><code>## 
## FALSE  TRUE 
##    29    69</code></pre>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">Chi.squared</th>
<th align="right">Paired.Wilcoxon</th>
<th align="right">Paired.t.test</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p-value</td>
<td align="right">5.33e-05</td>
<td align="right">4.5e-06</td>
<td align="right">0.15</td>
</tr>
</tbody>
</table>
<p>While the t-test is less conclusive, both the chi-squared and paired Wilcoxon tests strongly suggest that plasticity is higher in wild accessions than in domesticated ones. However, this result might be misleading, because the values for the Raphanus domesticates combine multiple different species, which may have resulted in depressed estimates of gene expression plasticity. For a more robust analysis, we need to run the analysis for each Raphanus domesticate individually.</p>
<table>
<thead>
<tr class="header">
<th align="left">species</th>
<th align="right">up_wild</th>
<th align="right">up_dom</th>
<th align="right">chisq</th>
<th align="right">t</th>
<th align="right">wilcox</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Raphanus raphanistrum munra</td>
<td align="right">59</td>
<td align="right">39</td>
<td align="right">0.043400</td>
<td align="right">0.0484</td>
<td align="right">0.0008655</td>
</tr>
<tr class="even">
<td align="left">Raphanus sativus</td>
<td align="right">67</td>
<td align="right">31</td>
<td align="right">0.000276</td>
<td align="right">0.0928</td>
<td align="right">0.0005876</td>
</tr>
<tr class="odd">
<td align="left">Raphanus sativus var. caudatus</td>
<td align="right">66</td>
<td align="right">32</td>
<td align="right">0.000594</td>
<td align="right">0.7170</td>
<td align="right">0.0002253</td>
</tr>
</tbody>
</table>
<p>Again, while the t-test results are less conclusive, per the Wilcox and Chi-squared tests it’s clear that for interaction genes plasticity is higher overall in the wild progenitor than in modern domesticates. We must therefore conclude that domestication has selected for decreased gene expression plasticity in radishes.</p>
</div>
<div id="brassica-phenotypic-expression" class="section level3">
<h3>Brassica phenotypic expression</h3>
<p>Here, we perform a similar analysis to above but using phenotypic traits instead of genes as the response variable. In each case we use a gaussian linear model, except for leaf number which is analysed using a Poisson model since it’s a discrete trait. Here we include random effects, so the model design is:</p>
<p><span class="math display">\[expression \sim domestication + treatment + domestication*treatment + (1|population) + (1|parental\ standardization\ status)\]</span></p>
<p>Here we plot the p-values and coefficients (in parentheses) for each independent effect for each trait. FDR correction is applied across each row and response variables are scaled prior to analysis, again with the exception of those modeled using a Poisson distribution.</p>
<p><img src="Analysis/Phenotypic_analysis/Images/phenotypic_lmer_pvals_brassica_subset.png" width="2362" /></p>
<p>Our primary interest is in the bottom row, since significant terms here will indicate traits that exhibit differential plasticity in domesticated vs wild rapa. However, none of these interaction terms are significant. Nor, indeed, are any traits significantly different between wild and domesticated rapa overall, which is very surprising! However, we’re forced to conclude that none of the traits examined appear to have shown a shift in plasticity in the course of Brassica rapa domestication.</p>
</div>
<div id="raphanus-phenotypic-expression" class="section level3">
<h3>Raphanus phenotypic expression</h3>
<p><img src="Analysis/Phenotypic_analysis/Images/phenotypic_lmer_pvals_raphanus_subset.png" width="2362" /></p>
<p>Again, we reach the unfortunate conclusion that there seem to be no significant terms in any of the models. Might we find some if we subset by the separate Raphanus domesticates, rather than combining them all into one? Below we plot the interaction terms only for each comparison:</p>
<p><img src="Analysis/Phenotypic_analysis/Images/phenotypic_lmer_pvals_raph_doms_perspecies.png" width="2362" /></p>
<p>To the contrary, subsetting by domesticate further erases any traces of interaction significance that we might have found for Raphanus (not least because we have to apply FDR correction over 3 x 13 = 39 comparisons here).</p>
<p>We must therefore conclude that domestication hasn’t selected for greeater phenotypic plasticity in domesticates relative to their wild progenitors.</p>
</div>
</div>
<div id="work-package-2-comparing-wild-progenitors-to-never-domesticated-wild-species" class="section level2">
<h2>Work package 2: Comparing wild progenitors to never-domesticated wild species</h2>
<p>We now wish to ask a different question: do ‘progenitor species’ (i.e. those that were the original targets of human selection) differ systematically in their patterns of gene and phenotypic expression from wild species that were never domesticated? We are particularly interested in knowing whether progenitor species express greater phenotypic plasticity than their never-domesticated relatives, since this could suggest that earty human farmers either intentionally or inadvertently targeted species that were particularly receptive to the benefits of cultivation. Unfortunately, for Raphanus we have no data for non-domesticated wild species, so this analysis will be restricted to Brassica, comparing wild Brassica rapa to a number of never-domesticated Brassicas.</p>
<div id="gene-expression" class="section level3">
<h3>Gene expression</h3>
<p>We’ll begin with the results of a DESeq2 model as before, except that instead of comparing wild to domesticated, we’ll compare progenitors (wild Brassica rapa) to non-progenitors (all other wild Brassicas).</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">DEGs</th>
<th align="right">GO_terms</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Domesticated_bias</td>
<td align="right">8710</td>
<td align="right">53</td>
</tr>
<tr class="even">
<td>Wild_bias</td>
<td align="right">9520</td>
<td align="right">75</td>
</tr>
<tr class="odd">
<td>Unstressed_bias</td>
<td align="right">9</td>
<td align="right">40</td>
</tr>
<tr class="even">
<td>Stressed_bias</td>
<td align="right">11</td>
<td align="right">10</td>
</tr>
<tr class="odd">
<td>Interaction</td>
<td align="right">63</td>
<td align="right">30</td>
</tr>
</tbody>
</table>
<p>The results here are similar to those we generated when comparing the wild and domesticated Brassica rapas: many genes dividing the different species, few differing with treatment, and a mid-sized number with interaction terms. Again, our primary interest is in these interaction genes, since these are those that show differential plasticity between the progenitor and its never-domesticated relatives. Plotting the 12 genes with the lowest p-vlaues for this comparison:</p>
<p><img src="Analysis/RNAseq/Images/brassica_interaction_deg_plots_wilds.png" width="2362" /></p>
<p>Eyeballing these plots, there’s surprisingly little difference in the reaction norms for progenitor and non-progenitor accessions, so it seems likely that the effect sizes we’re observing here are quite low despite the small p-values. Nonetheless, we can check whether there’s an overall trend in terms of increases/decreases in plasticity between the two groups:</p>
<pre><code>## [1] &quot;Interaction genes with higher plasticity in progenitor: 36&quot;</code></pre>
<pre><code>## [1] &quot;Interaction genes with higher plasticity in other wilds: 27&quot;</code></pre>
<pre><code>## [1] &quot;Plasticity higher in progenitor vs other wilds?:&quot;</code></pre>
<pre><code>## 
## FALSE  TRUE 
##    27    36</code></pre>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">Chi.squared</th>
<th align="right">Paired.Wilcoxon</th>
<th align="right">Paired.t.test</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>p-value</td>
<td align="right">0.274</td>
<td align="right">0.221</td>
<td align="right">0.786</td>
</tr>
</tbody>
</table>
<p>We find no effect here, but recall that this analysis combines multiple different wild Brassicas, which could depress any effect sizes we find. To be certain, we now subset by wild Brassica species:</p>
<table>
<thead>
<tr class="header">
<th align="left">species</th>
<th align="right">up_otherwild</th>
<th align="right">up_progenitor</th>
<th align="right">chisq</th>
<th align="right">t</th>
<th align="right">wilcox</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Brassica montana</td>
<td align="right">35</td>
<td align="right">28</td>
<td align="right">0.3780</td>
<td align="right">0.333</td>
<td align="right">0.1311542</td>
</tr>
<tr class="even">
<td align="left">Brassica cretica</td>
<td align="right">31</td>
<td align="right">32</td>
<td align="right">0.9000</td>
<td align="right">0.809</td>
<td align="right">0.6538502</td>
</tr>
<tr class="odd">
<td align="left">Brassica macrocarpa</td>
<td align="right">32</td>
<td align="right">31</td>
<td align="right">0.9000</td>
<td align="right">0.658</td>
<td align="right">0.2543378</td>
</tr>
<tr class="even">
<td align="left">Brassica rupestris</td>
<td align="right">25</td>
<td align="right">38</td>
<td align="right">0.1010</td>
<td align="right">0.894</td>
<td align="right">0.2216964</td>
</tr>
<tr class="odd">
<td align="left">Brassica incana</td>
<td align="right">24</td>
<td align="right">39</td>
<td align="right">0.0588</td>
<td align="right">0.431</td>
<td align="right">0.3571543</td>
</tr>
<tr class="even">
<td align="left">Brassica villosa</td>
<td align="right">29</td>
<td align="right">34</td>
<td align="right">0.5290</td>
<td align="right">0.848</td>
<td align="right">0.9918065</td>
</tr>
</tbody>
</table>
<p>Even when subsetting, it’s clear that there is no overall trend for the Brassica progenitor species to have higher or lower gene expression plasticity in interaction genes when comapred to never-domesticated Brassicas.</p>
</div>
<div id="phenotypic-expression" class="section level3">
<h3>Phenotypic expression</h3>
<p>What about phenotypes- does wild B rapa exhibit increased phenotypic plasticity relative to its never-domesticated relatives, upon which human ancestors might have selected?</p>
<p><img src="Analysis/Phenotypic_analysis/Images/phenotypic_lmer_pvals_brassica_wilds.png" width="2362" /></p>
<p>Here we see a more interesting set of results: many traits differ between wild Rapa and the other wild species, and a few traits have near-significant interaction terms: height, above-ground dry weight, and root dry weight. Perhaps these interaction terms will become more clearly significant (or non-significant) if we subset the never-domesticated wilds by species instead of combining them all into one?</p>
<p><img src="Analysis/Phenotypic_analysis/Images/phenotypic_lmer_pvals_brass_wilds_perspecies.png" width="2362" /></p>
<p>Alas, to the contrary: when we subset, we lose any hint of significance in the interaction terms. Note that these values will once again have been hit hard by FDR correction, since we have to correct for 13x6=78 comparisons. Nonetheless, we’re forced to conclude that Brassica rapa was not selected for its uniquly high plasticity at the level of either gene expression or phenotypic expression.</p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
