#need counts matrix with cols combining both gen4 and gen2 samples

metadata.brass.comp = mutate(metadata.brass,accession=substr(label,5,11))
metadata.brass.gen4.comp = metadata.brass.gen4 %>% subset(sample!="RN003") %>%
  mutate(accession=substr(label,3,9))
#combine matching samples
metadata.brass.comp.match = subset(metadata.brass.comp,accession %in% metadata.brass.gen4.comp$accession)
metadata.brass.combined = rbind(metadata.brass.comp.match,metadata.brass.gen4.comp)
#get relevant rnaseq data
brass.gene.counts.comp = select(brass.gene.counts, metadata.brass.comp.match$sample)
brass.gene.counts.gen4.comp = select(brass.gene.counts.gen4, metadata.brass.gen4.comp$sample)
#combined RNAseq data
brass.gene.counts.combined = cbind(brass.gene.counts.comp,brass.gene.counts.gen4.comp)
#check conformity
table(colnames(brass.gene.counts.combined) == metadata.brass.combined$sample)

#filter by expression
brass.gene.counts.combined.clean = brass.gene.counts.combined[(which(rowMeans(brass.gene.counts.combined)>=1)),]
print(paste0(nrow(brass.gene.counts.combined)-nrow(brass.gene.counts.combined.clean),"/",nrow(brass.gene.counts.combined)," Brassica genes filtered due to very low expression."))
#an additional filter- remove genes that have too many 0 counts
print(paste0("Removing an additional ",nrow(brass.gene.counts.combined.clean[rowSums(brass.gene.counts.combined.clean >= 5) < 3,])," Brassica genes with many low counts."))
brass.gene.counts.combined.clean = brass.gene.counts.combined.clean[rowSums(brass.gene.counts.combined.clean >= 5) >= 3,]

#now run model
dds.gene = DESeqDataSetFromMatrix(countData = brass.gene.counts.combined.clean,
                                  colData = metadata.brass.combined,
                                  design = as.formula(~treatment+generation+treatment*generation))

# Run the default analysis for DESeq2
#NA p-values are generated by 0 counts and outliers calculated by Cook's distance.
dds.gene.deg = DESeq(dds.gene, fitType = "parametric", betaPrior = FALSE)
# Check for outliers: none are apparent
print("Check for gene expression outliers")
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds.gene.deg)[["cooks"]]), range=0, las=2)
boxplot(log10(assays(dds.gene.deg)[["counts"]]), range=0, las=2)
#Also plot the dispersion extimates to make sure they look fine
plotDispEsts(dds.gene.deg)

#results: interaction
degs.brass.combined.interaction = results(dds.gene.deg, 
                                 name="treatmentControl.generation4",     
                                 alpha = 0.05,
                                 lfcThreshold = log2(1))
summary(degs.brass.combined.interaction) #12 degs
degs.brass.combined.interaction.ids = rownames(subset(degs.brass.combined.interaction, padj<=0.05))
degs.brass.combined.interaction.N = length(degs.brass.combined.interaction.ids)

##plot interaction terms
#for the interaction terms, we also plot the output to understand what exactly is going on
#get the 12 terms with lowest adjusted p value
degs.brass.combined.interaction.signif = subset(degs.brass.combined.interaction, padj<0.05)
interestgenes = row.names(degs.brass.combined.interaction.signif)[order(degs.brass.combined.interaction.signif$log2FoldChange,decreasing = T)[1:12]]
#for each gene, extract the counts for plotting and label by gene
for(i in 1:length(interestgenes)){
  #if first element, instantiate frame, otherwise rbind to frame
  if(i==1) {
    brass.intplotdata = plotCounts(dds.gene.deg, gene=interestgenes[i], intgroup=c("generation","treatment"), returnData = T)
    brass.intplotdata$gene = interestgenes[i]
  } else {
    addrows = data.frame(plotCounts(dds.gene.deg, gene=interestgenes[i], intgroup=c("generation","treatment"), returnData = T), 
                         gene = interestgenes[i])
    brass.intplotdata = rbind(brass.intplotdata, addrows)
  }
}

#plot with ggplot facet wrap
brass.intplot= ggplot(brass.intplotdata, aes(x = treatment, y = count)) +
  stat_summary(aes(group = generation), fun = mean, geom = "path") +
  stat_summary(aes(color = generation), fun.data = mean_cl_boot, geom = "errorbar", width = 0.1) +
  stat_summary(aes(color = generation), fun = mean, geom = "point", size = 4) +
  geom_point(aes(color = generation), size = 2) +
  scale_color_manual(values = brewer.pal(6,"Set3")[c(5,6)]) +
  facet_wrap(~gene, scales = "free")
brass.intplot
#save plot
ggsave(brass.intplot, 
       filename = "brass_gen2_gen4_interaction_deg_plots.png",
       device = "png", path = "Analysis/RNAseq/Images/",
       width =  40, height = 25, units = "cm")

#doesn't look like a consistent difference, but let's try
#set dplyr options to avoid excessively verbose output
options(dplyr.summarise.inform=F)
#loop across interaction genes, checking whether plasticity is higher in gen2 or gen4
for(i in 1:length(degs.brass.combined.interaction.ids)){
  #for each gene, summarise count values by generation and treatment
  intgene.data = plotCounts(dds.gene.deg, gene=degs.brass.combinedinteraction.ids[i], intgroup=c("generation","treatment"), returnData = T)
  intgene.counts = intgene.data %>% group_by(generation,treatment) %>% 
    summarise(mean = mean(count))
  #get absolute difference in count means
  intgene.diff.frame = data.frame(gene = degs.brass.interaction.ids[i],
                                  gen4 = abs(diff(subset(intgene.counts,generation == 2)$mean)),
                                  gen2 = abs(diff(subset(intgene.counts,generation == 4)$mean)))
  #check whether absolute difference in count means is higher for gen4 or for cultivated
  intgene.diff.frame = mutate(intgene.diff.frame, gen4.higher = (gen4>gen2))
  #collate output
  if(i==1){intgene.diff.out = intgene.diff.frame}else{intgene.diff.out = rbind(intgene.diff.out,intgene.diff.frame)}
}

#run a chisq test here checking whether higher gen4 or gen2 is statistically overrepresented
table(intgene.diff.out$gen4.higher)
brass.gen2.gen4.inter.chisq = chisq.test(table(intgene.diff.out$gen4.higher))
#run a paired t-test comparing fold changes in gen4 or gen2esticated
brass.gen2.gen4.inter.t.test = t.test(intgene.diff.out$gen4,intgene.diff.out$gen2,paired=T)
brass.gen2.gen4.inter.wilcox = wilcox.test(intgene.diff.out$gen4,intgene.diff.out$gen2,paired=T)

#save output
write.csv(intgene.diff.out,file = "Analysis/RNAseq/brassica_gen2_gen4_interaction_gene_directionality_table.csv")
save(brass.gen2.gen4.inter.t.test,file = "Analysis/RNAseq/brassica_gen2_gen4_interaction_gene_directionality_ttest.R")
save(brass.gen2.gen4.inter.chisq,file = "Analysis/RNAseq/brassica_gen2_gen4_interaction_gene_directionality_chisq.R")
save(brass.gen2.gen4.inter.wilcox,file = "Analysis/RNAseq/brassica_gen2_gen4_interaction_gene_directionality_wilcox.R")



#####Now for Raphanus####
#need counts matrix with cols combining both gen4 and gen2 samples

metadata.raph.comp = mutate(metadata.raph,accession=substr(label,5,11))
metadata.raph.gen4.comp = metadata.raph.gen4 %>% subset(sample!="RN003") %>%
  mutate(accession=substr(label,3,9))
#combine matching samples
metadata.raph.comp.match = subset(metadata.raph.comp,accession %in% metadata.raph.gen4.comp$accession)
metadata.raph.combined = rbind(metadata.raph.comp.match,metadata.raph.gen4.comp)
#get relevant rnaseq data
raph.gene.counts.comp = select(raph.gene.counts, metadata.raph.comp.match$sample)
raph.gene.counts.gen4.comp = select(raph.gene.counts.gen4, metadata.raph.gen4.comp$sample)
#combined RNAseq data
raph.gene.counts.combined = cbind(raph.gene.counts.comp,raph.gene.counts.gen4.comp)
#check conformity
table(colnames(raph.gene.counts.combined) == metadata.raph.combined$sample)

#filter by expression
raph.gene.counts.combined.clean = raph.gene.counts.combined[(which(rowMeans(raph.gene.counts.combined)>=1)),]
print(paste0(nrow(raph.gene.counts.combined)-nrow(raph.gene.counts.combined.clean),"/",nrow(raph.gene.counts.combined)," raphica genes filtered due to very low expression."))
#an additional filter- remove genes that have too many 0 counts
print(paste0("Removing an additional ",nrow(raph.gene.counts.combined.clean[rowSums(raph.gene.counts.combined.clean >= 5) < 3,])," raphica genes with many low counts."))
raph.gene.counts.combined.clean = raph.gene.counts.combined.clean[rowSums(raph.gene.counts.combined.clean >= 5) >= 3,]

#now run model
dds.gene = DESeqDataSetFromMatrix(countData = raph.gene.counts.combined.clean,
                                  colData = metadata.raph.combined,
                                  design = as.formula(~treatment+generation+treatment*generation))

# Run the default analysis for DESeq2
#NA p-values are generated by 0 counts and outliers calculated by Cook's distance.
dds.gene.deg = DESeq(dds.gene, fitType = "parametric", betaPrior = FALSE)
# Check for outliers: none are apparent
print("Check for gene expression outliers")
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds.gene.deg)[["cooks"]]), range=0, las=2)
boxplot(log10(assays(dds.gene.deg)[["counts"]]), range=0, las=2)
#Also plot the dispersion extimates to make sure they look fine
plotDispEsts(dds.gene.deg)

#results: interaction
degs.raph.combined.interaction = results(dds.gene.deg, 
                                          name="treatmentControl.generation4",     
                                          alpha = 0.05,
                                          lfcThreshold = log2(1))
summary(degs.raph.combined.interaction) #52 degs
degs.raph.combined.interaction.ids = rownames(subset(degs.raph.combined.interaction, padj<=0.05))
degs.raph.combined.interaction.N = length(degs.raph.combined.interaction.ids)

##plot interaction terms
#for the interaction terms, we also plot the output to understand what exactly is going on
#get the 12 terms with lowest adjusted p value
degs.raph.combined.interaction.signif = subset(degs.raph.combined.interaction, padj<0.05)
interestgenes = row.names(degs.raph.combined.interaction.signif)[order(degs.raph.combined.interaction.signif$log2FoldChange,decreasing = T)[1:12]]
#for each gene, extract the counts for plotting and label by gene
for(i in 1:length(interestgenes)){
  #if first element, instantiate frame, otherwise rbind to frame
  if(i==1) {
    raph.intplotdata = plotCounts(dds.gene.deg, gene=interestgenes[i], intgroup=c("generation","treatment"), returnData = T)
    raph.intplotdata$gene = interestgenes[i]
  } else {
    addrows = data.frame(plotCounts(dds.gene.deg, gene=interestgenes[i], intgroup=c("generation","treatment"), returnData = T), 
                         gene = interestgenes[i])
    raph.intplotdata = rbind(raph.intplotdata, addrows)
  }
}

#plot with ggplot facet wrap
raph.intplot= ggplot(raph.intplotdata, aes(x = treatment, y = count)) +
  stat_summary(aes(group = generation), fun = mean, geom = "path") +
  stat_summary(aes(color = generation), fun.data = mean_cl_boot, geom = "errorbar", width = 0.1) +
  stat_summary(aes(color = generation), fun = mean, geom = "point", size = 4) +
  geom_point(aes(color = generation), size = 2) +
  scale_color_manual(values = brewer.pal(6,"Set3")[c(5,6)]) +
  facet_wrap(~gene, scales = "free")
raph.intplot
#save plot
ggsave(raph.intplot, 
       filename = "raph_gen2_gen4_interaction_deg_plots.png",
       device = "png", path = "Analysis/RNAseq/Images/",
       width =  40, height = 25, units = "cm")

#here it looks like reaction norms might be a little higher on avergae in gen 2
#set dplyr options to avoid excessively verbose output
options(dplyr.summarise.inform=F)
#loop across interaction genes, checking whether plasticity is higher in gen2 or gen4
for(i in 1:length(degs.raph.combined.interaction.ids)){
  #for each gene, summarise count values by generation and treatment
  intgene.data = plotCounts(dds.gene.deg, gene=degs.raph.combined.interaction.ids[i], intgroup=c("generation","treatment"), returnData = T)
  intgene.counts = intgene.data %>% group_by(generation,treatment) %>% 
    summarise(mean = mean(count))
  #get absolute difference in count means
  intgene.diff.frame = data.frame(gene = degs.raph.interaction.ids[i],
                                  gen4 = abs(diff(subset(intgene.counts,generation == 2)$mean)),
                                  gen2 = abs(diff(subset(intgene.counts,generation == 4)$mean)))
  #check whether absolute difference in count means is higher for gen4 or for cultivated
  intgene.diff.frame = mutate(intgene.diff.frame, gen4.higher = (gen4>gen2))
  #collate output
  if(i==1){intgene.diff.out = intgene.diff.frame}else{intgene.diff.out = rbind(intgene.diff.out,intgene.diff.frame)}
}

#run a chisq test here checking whether higher gen4 or gen2 is statistically overrepresented
table(intgene.diff.out$gen4.higher)
raph.gen2.gen4.inter.chisq = chisq.test(table(intgene.diff.out$gen4.higher))
#run a paired t-test comparing fold changes in gen4 or gen2esticated
raph.gen2.gen4.inter.t.test = t.test(intgene.diff.out$gen4,intgene.diff.out$gen2,paired=T)
raph.gen2.gen4.inter.wilcox = wilcox.test(intgene.diff.out$gen4,intgene.diff.out$gen2,paired=T)

#save output
write.csv(intgene.diff.out,file = "Analysis/RNAseq/raphanus_gen2_gen4_interaction_gene_directionality_table.csv")
save(raph.gen2.gen4.inter.t.test,file = "Analysis/RNAseq/raphanus_gen2_gen4_interaction_gene_directionality_ttest.R")
save(raph.gen2.gen4.inter.chisq,file = "Analysis/RNAseq/raphanus_gen2_gen4_interaction_gene_directionality_chisq.R")
save(raph.gen2.gen4.inter.wilcox,file = "Analysis/RNAseq/raphanus_gen2_gen4_interaction_gene_directionality_wilcox.R")
