library("DESeq2")
library("tidyverse")

setwd("/home/benjamin/Documents/Brassicas_repo")

#import functions
source("Functions/expression_heatmap.R")

#pull relevant metadata
data.meta = read.csv("/home/benjamin/Documents/Brassicas_repo/Data/RNAseq/RNASeq_sample_info.csv") %>%
  mutate(Domesticated = ifelse(Wild..Domesticated=="Wild","Wild","Cultivated")) %>%
  mutate(Environment = ifelse(Environment=="wheat competition","Wheat","Control"),
         Parental.effects.status = ifelse(Parental.effects.status == "'\"standardised\"","Standardised","Unstandardised")) %>%
  select(c("RNAseq.sample.name","Species","Parental.effects.status","Environment","Domesticated")) %>%
  'colnames<-'(c("sample","species","parental.effects","treatment","domesticated")) %>%
  mutate_all(as.factor)

data.meta = mutate(data.meta, 
                   treatment = fct_relevel(treatment, c("Control","Wheat")),
                   domesticated = fct_relevel(domesticated, c("Wild","Cultivated")))


#subset by genus
metadata.brass = subset(data.meta, substr(species,1,3)=="Bra")
metadata.raph = subset(data.meta, substr(species,1,3)=="Rap")

#pull rnaseq data
raph.gene.counts = read.csv("/home/benjamin/Documents/Brassicas_repo/Data/RNAseq/raph.gene.counts.csv", row.names = 1)

#filter by expression
raph.gene.counts.clean = raph.gene.counts[(which(rowMeans(raph.gene.counts)>=1)),]
paste0(nrow(raph.gene.counts)-nrow(raph.gene.counts.clean),"/",nrow(raph.gene.counts)," Raphanus genes filtered due to very low expression.")

#an additional filter- remove genes that have too many 0 counts
paste0("Removing an additional ",nrow(raph.gene.counts.clean[rowSums(raph.gene.counts.clean >= 5) < 3,])," genes with many low counts.")
raph.gene.counts.clean = raph.gene.counts.clean[rowSums(raph.gene.counts.clean >= 5) >= 3,]

#check that metadata and count matrices conform
table(metadata.raph$sample %in% colnames(raph.gene.counts.clean))
raph.gene.counts.clean = raph.gene.counts.clean[,as.character(metadata.raph$sample)]

#some basic QC: heatmaps and pca
# heatMap = expression.heatmap(countdata = sample_n(raph.gene.counts.clean,5000), 
#                              data.phenotype = metadata.raph,
#                              labels = c("parental.effects",
#                                         "treatment",
#                                         "domesticated"),
#                              pass_on = F,
#                              ID_var = "sample")

# # log-transform with a pseudocount
# pca.counts = log2(raph.gene.counts.clean+1)
# #create pca object
# data.pca = prcomp(t(pca.counts))
# #extract PC data
# percent.var = (data.pca$sdev^2 / sum(data.pca$sdev^2))
# pca.out = list(values = data.frame(data.pca$x),
#                percent.var = percent.var)
# #connect to phenotypic data
# ggpcadata = pca.out$values %>% 
#   rownames_to_column(var = "sample") %>%
#   left_join(metadata.raph,
#             by = "sample")
# #plot
# ggplot(ggpcadata, aes(x = PC1, y = PC2, color = domesticated, shape = treatment, label = sample)) +
#   geom_point(size = 5, position = position_jitter(width = 0.5,height=0.5)) +
#   #geom_text(vjust = -1) +
#   xlab(paste0("PC",1,": ",signif(pca.out$percent.var[1]*100, 3),"%")) +
#   ylab(paste0("PC",2,": ",signif(pca.out$percent.var[2]*100, 3),"%")) +
#   theme_bw() +
#   # scale_color_manual(name = "Treatment",
#   #                    values = brewer.pal(7, "Paired")) +
#   scale_shape_manual(name = "Treatment",
#                      values = c(8,15:20)) +
#   theme(panel.grid = element_line(color = "grey95"),
#         legend.title = element_text(face = "bold"),
#         axis.text.x = element_text(size = 11),
#         axis.text.y = element_text(size = 11),
#         axis.title = element_text(face = "bold", size =12))




#next step is to check for genes with parental effects and exclude these
dds.parental = DESeqDataSetFromMatrix(countData = raph.gene.counts.clean,
                                      colData = metadata.raph,
                                      design = as.formula(~parental.effects))
dds.parental.deg = DESeq(dds.parental, fitType = "local", betaPrior = FALSE)
parental.degs = results(dds.parental.deg)
#very few genes have even a hint of parental effect, so let's just drop these
table(parental.degs$padj<0.1)
raph.gene.counts.clean = raph.gene.counts.clean[rownames(subset(parental.degs, padj>=0.1)),]

#now run proper model
dds.gene = DESeqDataSetFromMatrix(countData = raph.gene.counts.clean,
                                  colData = metadata.raph,
                                  design = as.formula(~treatment+domesticated+treatment*domesticated))

# Run the default analysis for DESeq2 and generate results table. NA p-values are generated by 0 counts and outliers calculated by Cook's distance.
dds.gene.deg = DESeq(dds.gene, fitType = "local", betaPrior = FALSE)
# Check for outliers: none are apparent
print("Check for gene expression outliers")
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds.gene.deg)[["cooks"]]), range=0, las=2)

#Also plot the dispersion extimates to make sure they look fine
plotDispEsts(dds.gene.deg)

#check effectof a parametric rather than local fit
parametric.gene.deg = DESeq(dds.gene, fitType = "parametric", betaPrior = FALSE)
plotDispEsts(parametric.gene.deg) 
# parametric seems to follow the data better so let's go with that
dds.gene.deg = parametric.gene.deg

resultsNames(dds.gene.deg)

#wheat vs control
comparison = results(dds.gene.deg, 
                     name="treatment_Wheat_vs_Control",     
                     alpha = 0.05,
                     lfcThreshold = log2(2))
summary(comparison)

#cultivated vs wild
comparison = results(dds.gene.deg, 
                     name="domesticated_Cultivated_vs_Wild",     
                     alpha = 0.05,
                     lfcThreshold = log2(2))
summary(comparison)

#cultivated vs wild
comparison = results(dds.gene.deg, 
                     name="treatmentWheat.domesticatedCultivated",     
                     alpha = 0.05,
                     lfcThreshold = log2(2))
summary(comparison)

